### PR TITLE
Add torch and tf A2C

### LINKED ISSUE #547: Implement A2C

### PR DESCRIPTION
**Pull request by maliesa96**
closes #547

### CODE REVIEW

File: tests/garage/torch/algos/test_a2c.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class TestA2C:
+    """Test class for A2C."""
+
+    @pytest.mark.mujoco
+    def test_a2c_pendulum(self):
+        deterministic.set_seed(0)
+
+        env = GymEnv(gym.make('InvertedDoublePendulum-v2'))
```
Comment (ryanjulian) [inline tests/garage/torch/algos/test_a2c.py:23]:
```suggestion
        env = GymEnv('InvertedDoublePendulum-v2')
```

File: examples/tf/a2c_pendulum.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+from garage.tf.policies import GaussianMLPPolicy
+
+
+@wrap_experiment
+def a2c_pendulum(ctxt=None, seed=1):
+    """Train A2C with InvertedDoublePendulum-v2 environment.
+
+    Args:
+        ctxt (garage.experiment.ExperimentContext): The experiment
+            configuration used by LocalRunner to create the snapshotter.
```
Comment (ryanjulian) [inline examples/tf/a2c_pendulum.py:27]:
```suggestion
        ctxt (ExperimentContext): The experiment configuration used by LocalRunner to create the snapshotter.
```

File: src/garage/tf/algos/a2c.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    """A2C paper: https://arxiv.org/abs/1602.01783.
+
+    A2C is the synchronous variant of the A3C algorithm described
+    in the paper. The primary difference between A2C and A3C is that
+    the former uses one policy and baseline network and
+    synchronously updates these networks after all workers complete
+    the sampling step.
+
+    Args:
+        env_spec (garage.envs.EnvSpec): Environment specification.
```
Comment (ryanjulian) [inline src/garage/tf/algos/a2c.py:15]:
```suggestion
        env_spec (EnvSpec): Environment specification.
```

File: src/garage/torch/algos/a2c.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    """A2C paper: https://arxiv.org/abs/1602.01783.
+
+    A2C is the synchronous variant of the A3C algorithm described
+    in the paper. The primary difference between A2C and A3C is that
+    the former uses one policy and baseline network and
+    synchronously updates these networks after all workers complete
+    the sampling step.
+
+    Args:
+        env_spec (garage.envs.EnvSpec): Environment specification.
```
Comment (ryanjulian) [inline src/garage/torch/algos/a2c.py:16]:
```suggestion
        env_spec (EnvSpec): Environment specification.
```

File: tests/garage/tf/algos/test_a2c.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+from tests.fixtures import snapshot_config, TfGraphTestCase
+
+
+class TestA2C(TfGraphTestCase):
+
+    @pytest.mark.large
+    def test_a2c_cartpole(self):
+        """Test A2C with CartPole-v1 environment."""
+        with LocalTFRunner(snapshot_config, sess=self.sess) as runner:
+            env = GymEnv(env_name='CartPole-v1')
```
Comment (ryanjulian) [inline tests/garage/tf/algos/test_a2c.py:19]:
```suggestion
            env = GymEnv('CartPole-v1')
```

File: src/garage/tf/algos/a2c.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    """
+
+    def __init__(self,
+                 env_spec,
+                 policy,
+                 baseline,
+                 scope=None,
+                 max_episode_length=500,
+                 discount=0.99,
+                 gae_lambda=1,
```
Comment (ryanjulian) [inline src/garage/tf/algos/a2c.py:59]: does the A2C paper allow for GAE discounting? if not, i would not allow this to be an option

Comment (maliesa96) [reply src/garage/tf/algos/a2c.py]: It does not, setting this to 1.

**Comment by maliesa96** [comment]
benchmark results (Mujoco1M):

tb dev:
torch https://tensorboard.dev/experiment/ujMcBpXpRpaaVZN1P915mg/#scalars
tf https://tensorboard.dev/experiment/BYoQdKP4Q0irxnbvFzaGbw/

HalfCheetah-v2
Hopper-v2
InvertedDoublePendulum-v2
InvertedPendulum-v2
Reacher-v2
Swimmer-v2
Walker2d-v2

**Comment by ryanjulian** [comment]
1. How do you explain the variance difference between openai/baselines and garage? It seems that it might be as simple as there are very few data points for baselines, in which case you may need to make openai/baselines log more often
2. Are you sure you are using the same policy and value function (baseline) architecture across all implementations? Are all the hyperparameters the same? This has been a sticking point when comparing the openai/baselines in the past.
3. Not all of these seem converged -- it seems you might need to run them out to 3M time steps

**Comment by maliesa96** [comment]
Closed as we're prioritizing DQN and its variants over A2C/A3C.

**Pull request closed without merging**